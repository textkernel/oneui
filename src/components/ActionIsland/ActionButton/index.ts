export { ActionButton, type ActionButtonProps } from './ActionButton';
