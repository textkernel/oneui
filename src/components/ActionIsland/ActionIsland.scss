@keyframes islandAnimation {
    0% {
        transform: translateX(-50%) translateY(100%);
        opacity: 0;
    }
    100% {
        transform: translateX(-50%) translateY(0);
        opacity: 1;
    }
}

.ActionIsland {
    width: fit-content;
    padding: var(--space-50) var(--space-100) var(--space-50) var(--space-200);
    margin-bottom: var(--space-200);
    position: fixed;
    bottom: 0;
    border-radius: var(--space-100);
    background-color: var(--color-background);
    display: flex;
    align-items: center;
    gap: var(--space-100);
    z-index: 1000;
    left: 50%;
    transform: translateX(-50%);
    animation: islandAnimation 300ms ease-out;
    white-space: nowrap;
    overflow: hidden;

    &__label {
        padding-right: var(--space-100);
        border-right: 1px solid var(--color-border-disabled);
        text-wrap: nowrap;
    }
}
