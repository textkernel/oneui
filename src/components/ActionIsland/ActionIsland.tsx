import * as React from 'react';
import Close from '@material-design-icons/svg/round/close.svg';
import { bem } from '../../utils';
import { IconButton } from '../Buttons/IconButton/IconButton';
import { Text } from '../Text';
import styles from './ActionIsland.scss';
import { ActionButton } from './ActionButton';
import { Tooltip } from '../Tooltip/Tooltip';
import { ActionButtonProps } from './ActionButton/ActionButton';

export interface Props extends Omit<React.HTMLAttributes<HTMLDivElement>, 'children'> {
    actionButtons: ActionButtonProps[];
    /** Determines if the action island is shown. */
    isShown: boolean;
    /** Callback function triggered when the close button is clicked. */
    onClose: () => void;
    /** Counter part or prefix of the label. */
    size: React.ReactNode;
    /** Main label to be displayed */
    label: React.ReactNode;
    /** Label for the "More" button */
    moreButtonLabel?: string;
    /** Close button label name for ARIA labelling */
    closeButtonLabel?: string;
    /** Tooltip content for the close button, which appears on hover */
    closeButtonTooltip?: string;
}

const { block, elem } = bem('ActionIsland', styles);

export const ActionIsland: React.FC<Props> = ({
    actionButtons,
    isShown,
    onClose,
    size,
    label,
    moreButtonLabel,
    closeButtonLabel,
    closeButtonTooltip,
    ...rest
}) => {
    const visibleButtons = actionButtons.slice(0, 3);
    const overflowButtons = actionButtons.slice(3);

    return (
        isShown && (
            <div {...block()} {...rest}>
                <div {...elem('header')}>
                    <Text inline {...elem('label')}>
                        <Text inline isBold>
                            {size}
                        </Text>{' '}
                        {label}
                    </Text>
                </div>
                <div {...elem('actionsContainer')}>
                    {visibleButtons.map((button) => (
                        <ActionButton key={`${button.label}`} {...button} />
                    ))}
                    {overflowButtons.length > 0 && (
                        <ActionButton label={moreButtonLabel} dropdownItems={overflowButtons} />
                    )}
                </div>
                <Tooltip placement="top" content={closeButtonTooltip}>
                    <IconButton
                        variant="ghost"
                        size="large"
                        onClick={onClose}
                        aria-label={closeButtonLabel}
                    >
                        <Close viewBox="0 0 24 24" />
                    </IconButton>
                </Tooltip>
            </div>
        )
    );
};

ActionIsland.displayName = 'ActionIsland';
