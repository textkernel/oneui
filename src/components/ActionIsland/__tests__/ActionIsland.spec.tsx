import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import { ActionIsland, Props } from '../ActionIsland';

const mockOnClose = jest.fn();

const defaultProps: Props = {
    actionButtons: [
        { label: 'Action 1', onClick: jest.fn() },
        { label: 'Action 2', onClick: jest.fn() },
        { label: 'Action 3', onClick: jest.fn() },
        { label: 'Overflow Action', onClick: jest.fn() },
    ],
    isShown: true,
    onClose: mockOnClose,
    size: 'Large',
    label: 'Test Label',
    moreButtonLabel: 'More',
    closeButtonLabel: 'Close button',
    closeButtonTooltip: 'Close Tooltip',
};

describe('ActionIsland', () => {
    it('renders correctly with all props', () => {
        const { container } = render(<ActionIsland {...defaultProps} />);
        expect(screen.getByText('Test Label')).toBeInTheDocument();
        expect(screen.getByText('Action 1')).toBeInTheDocument();
        expect(screen.getByText('More')).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it('handles close button click correctly', async () => {
        render(<ActionIsland {...defaultProps} />);
        const user = userEvent.setup();
        const closeButton = screen.getByRole('button', { name: /close/i });
        await user.click(closeButton);
        expect(mockOnClose).toHaveBeenCalled();
    });

    it('renders visible and overflow buttons correctly', () => {
        const { container } = render(<ActionIsland {...defaultProps} />);
        expect(screen.getByText('Action 1')).toBeInTheDocument();
        expect(screen.getByText('More')).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it('does not render when `isShown` is false', () => {
        const { container } = render(<ActionIsland {...defaultProps} isShown={false} />);
        expect(screen.queryByText('Test Label')).not.toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });
});
