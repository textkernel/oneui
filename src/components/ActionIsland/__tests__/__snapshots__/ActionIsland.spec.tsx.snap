// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionIsland does not render when \`isShown\` is false 1`] = `<div />`;

exports[`ActionIsland renders correctly with all props 1`] = `
<div>
  <div
    class="ActionIsland"
  >
    <div>
      <span
        class="ActionIsland__label"
      >
        <span
          class="Text--isBold"
        >
          Large
        </span>
         
        Test Label
      </span>
    </div>
    <div>
      <button
        class="Button Button--variant_ghost Button--context_primary"
        type="button"
      >
        Action 1
      </button>
      <button
        class="Button Button--variant_ghost Button--context_primary"
        type="button"
      >
        Action 2
      </button>
      <button
        class="Button Button--variant_ghost Button--context_primary"
        type="button"
      >
        Action 3
      </button>
      <button
        aria-expanded="false"
        aria-haspopup="menu"
        class="Button Button--variant_ghost Button--context_primary"
        data-state="closed"
        id="radix-:r0:"
        type="button"
      >
        More
        <svg
          class="ActionButton__icon"
          data-testid="default-icon"
          height="20px"
          viewBox="0 0 24 24"
          width="20px"
        />
      </button>
    </div>
    <button
      aria-label="Close button"
      class="Button Button--size_large Button--variant_ghost Button--context_secondary IconButton--size_large"
      type="button"
    >
      <svg
        class="IconButton__icon--size_large"
        data-testid="default-icon"
        viewBox="0 0 24 24"
      />
    </button>
  </div>
</div>
`;

exports[`ActionIsland renders visible and overflow buttons correctly 1`] = `
<div>
  <div
    class="ActionIsland"
  >
    <div>
      <span
        class="ActionIsland__label"
      >
        <span
          class="Text--isBold"
        >
          Large
        </span>
         
        Test Label
      </span>
    </div>
    <div>
      <button
        class="Button Button--variant_ghost Button--context_primary"
        type="button"
      >
        Action 1
      </button>
      <button
        class="Button Button--variant_ghost Button--context_primary"
        type="button"
      >
        Action 2
      </button>
      <button
        class="Button Button--variant_ghost Button--context_primary"
        type="button"
      >
        Action 3
      </button>
      <button
        aria-expanded="false"
        aria-haspopup="menu"
        class="Button Button--variant_ghost Button--context_primary"
        data-state="closed"
        id="radix-:r4:"
        type="button"
      >
        More
        <svg
          class="ActionButton__icon"
          data-testid="default-icon"
          height="20px"
          viewBox="0 0 24 24"
          width="20px"
        />
      </button>
    </div>
    <button
      aria-label="Close button"
      class="Button Button--size_large Button--variant_ghost Button--context_secondary IconButton--size_large"
      type="button"
    >
      <svg
        class="IconButton__icon--size_large"
        data-testid="default-icon"
        viewBox="0 0 24 24"
      />
    </button>
  </div>
</div>
`;
